src/cmd/manage_nsfs.js

```diff
@@ -822,6 +839,318 @@ function _validate_access_keys(argv) {
+                }, REQUEST_XATTR);
+
+                await fh.close(fs_context);
+            } catch (error) {
+                console.error(`failed to process ${entry}`, error);
+                if (error.code === 'ENOENT') {
+                    // Why would someone issue a delete for a entry after issueing
+                    // a restore? Nevermind, ignore this entry 
+                    state.failed_post.push(entry);
+                }
+            }
+
+            entry = await reader.nextline();
+        }
+
+        return unexpected_failed.filter(f => !state.failed_post.includes(f)).length === 0;
+    });
+}
+
+/**
+ * should_run_eeadm returns true if either the current timestamp exceeds
+ * timestamp (read from timestamp_file) + interval or if the free space
+ * on the filesystem is lesser than `free_disk_perc`
+ * 
+ * NOTE: This assumes that `NSFS_WAL_DIR` is on the same mount and uses
+ * that to determine the free space
+ * @param {nb.NativeFSContext} fs_context
+ * @param {number} interval interval in which the migration should run
+ * @param {number} free_threshold percentage of disk that needs to be free
+ * @param {string} timestamp_file name of the file which tracks timestamp
+ * @returns {Promise<boolean>}
+ */
+async function should_run_eeadm(fs_context, interval, free_threshold, timestamp_file) {
+    // 1. Run every X mins
+    try {
+        const { data } = await nb_native().fs.readFile(fs_context, path.join(config.NSFS_WAL_DIR, timestamp_file));
+        const lastrun = new Date(data.toString('utf-8'));
+
+        if (lastrun.getTime() + interval < Date.now()) return true;
+    } catch (error) {
+        console.error('failed to read last run timestamp:', error);
+        if (error.code === 'ENOENT') {
+            console.log('creating timestamp file');
+            await nb_native().fs.writeFile(
+                fs_context,
+                path.join(config.NSFS_WAL_DIR, timestamp_file),
+                Buffer.from(new Date().toISOString()),
+            );
+        }
+    }
+
+    // 2. Run if the free disk space is less than X percent
+    try {
+        const fsstat = await nb_native().fs.statfs(fs_context, config.NSFS_WAL_DIR);
+        const free = fsstat.bfree / fsstat.blocks;
+        if (free <= free_threshold) return true;
+    } catch (error) {
+        console.error('failed to run fsstat on', config.NSFS_WAL_DIR);
+    }
+
+    return false;
+}
+
+/**
+ * acquire_file_lock takes a lockfile name and optionally a mode and will
+ * create/open a file and acquire lock on it and return the lock file fd.
+ * @param {nb.NativeFSContext} fs_context 
+ * @param {string} lockfile 
+ * @param {'EXCLUSIVE' | 'SHARED'} mode 
+ * @returns {Promise<nb.NativeFile>}
+ */
+async function acquire_file_lock(fs_context, lockfile, mode = 'EXCLUSIVE') {
+    const lockfd = await nb_native().fs.open(fs_context, path.join(config.NSFS_WAL_DIR, lockfile), 'rw');
+    await lockfd.flock(fs_context, mode);
```

Comment:
if this fails we leak the open file. I would suggest to inline this into the lock_and_run function to avoid such an issue.